ui/config/config.js

```diff
@@ -0,0 +1,280 @@
+        console.error('Error parsing JSON from ' + filename + ': ' + exception);
+        continue;
+      }
+    }
+
+    let templates = UserConfig.optionTemplates;
+    for (let group in templates) {
+      this.contents[group] = this.contents[group] || [];
+      this.contents[group].push(templates[group]);
+    }
+
+    this.buildButterBar();
+
+    let container = document.getElementById('container');
+    this.buildUI(container, this.contents);
+  }
+
+  async saveConfigData() {
+    // TODO: rate limit this?
+    await callOverlayHandler({
+      call: 'cactbotSaveData',
+      overlay: 'options',
+      data: this.savedConfig,
+    });
+
+    document.getElementById('butter-margin').classList.remove('hidden');
+  }
+
+  buildButterBar() {
+    let container = document.getElementById('butter-bar');
+
+    let textDiv = document.createElement('div');
+    textDiv.classList.add('reload-text');
+    textDiv.innerText = this.translate(kReloadText);
+    container.appendChild(textDiv);
+
+    let buttonInput = document.createElement('input');
+    buttonInput.classList.add('reload-button');
+    buttonInput.type = 'button';
+    buttonInput.onclick = () => {
+      callOverlayHandler({ call: 'cactbotReloadOverlays' });
+    };
+    buttonInput.value = this.translate(kReloadButtonText);
+    container.appendChild(buttonInput);
+  }
+
+  buildOverlayGroup(container, group) {
+    let collapser = document.createElement('div');
+    collapser.classList.add('overlay-container');
+    container.appendChild(collapser);
+
+    let a = document.createElement('a');
+    a.name = group;
+    collapser.appendChild(a);
+
+    let header = document.createElement('div');
+    header.classList.add('overlay-header');
+    header.innerText = group;
+    a.appendChild(header);
+
+    let groupDiv = document.createElement('div');
+    groupDiv.classList.add('overlay-options');
+    collapser.appendChild(groupDiv);
+
+    a.onclick = (e) => {
+      a.parentNode.classList.toggle('collapsed');
+    };
+
+    return groupDiv;
+  }
+
+  buildUI(container, contents) {
+    for (let group in contents) {
+      let content = contents[group];
+      if (content.length == 0)
+        continue;
+
+      // For each overlay options template, build a section for it.
+      // Then iterate through all of its options and build ui for those options.
+      // Give each options template a chance to build special ui.
+      let groupDiv = this.buildOverlayGroup(container, group);
+      for (let i = 0; i < content.length; ++i) {
+        let options = content[i].options || [];
+        for (let j = 0; j < options.length; ++j) {
+          let opt = options[j];
+          if (!this.developerOptions && opt.debugOnly)
+            continue;
+          let buildFunc = {
+            checkbox: this.buildCheckbox,
+            select: this.buildSelect,
+            float: this.buildFloat,
+            integer: this.buildInteger,
+            directory: this.buildDirectory,
```

Comment:
Currently it looks like directory is unused; is the intent here for something like specifying custom sound files, or would that be an additional file/string option or is it a remnant of a user-specified user directory option?

Reply:
It's the remnant of a user-specified user directory option.

Today, I tried implementing a directory picker in C#, but didn't have any luck.  `FolderBrowserDialog`, even when I invoked it on the main act form, just didn't show anything when I used it.  `CommonOpenFileDialog` crashed, so I'm not sure what to do there.  @ngld suggested this approach, so I'll ask him for more suggestions.